lib/Arch/Sleigh/ControlFlowStructuring.cpp

```diff
@@ -0,0 +1,470 @@
+#include "ControlFlowStructuring.h"
+
+#include <glog/logging.h>
+
+namespace remill::sleigh {
+
+bool isVarnodeInConstantSpace(VarnodeData vnode) {
+  auto spc = vnode.getAddr().getSpace();
+  return spc->constant_space_index == spc->getIndex();
+}
+
+
+namespace {
+
+// variant casting taken c&ped.
+template <class... Args>
+struct variant_cast_proxy {
+  std::variant<Args...> v;
+
+  template <class... ToArgs>
+  operator std::variant<ToArgs...>() const {
+    return std::visit([](auto &&arg) -> std::variant<ToArgs...> { return arg; },
+                      v);
+  }
+};
+
+template <class... Args>
+auto variant_cast(const std::variant<Args...> &v)
+    -> variant_cast_proxy<Args...> {
+  return {v};
+}
+
+enum CoarseEffect { ABNORMAL, NORMAL };
+
+struct CoarseFlow {
+  CoarseEffect eff;
+  bool is_conditional;
+};
+
+enum CoarseCategory { CAT_NORMAL, CAT_ABNORMAL, CAT_CONDITIONAL_ABNORMAL };
+
+
+static std::optional<CoarseFlow>
+CoarseFlowFromControlFlowOp(const RemillPcodeOp &op, uint64_t next_pc) {
+  if (op.op == CPUI_CALL || op.op == CPUI_CALLIND || op.op == CPUI_BRANCHIND ||
+      op.op == CPUI_RETURN) {
+    return {{CoarseEffect::ABNORMAL, false}};
+  }
+
+  // either a branch or a cbranch
+
+  // figure out if this is a fallthrough, input 0 is the next target
+
+  auto is_conditional = op.op == CPUI_CBRANCH;
+  if (isVarnodeInConstantSpace(op.vars[0])) {
+    // this is an internal branch.. we cant handle that right now
+    return std::nullopt;
+  }
+
+  if (op.vars[0].offset == next_pc) {
+    return {{CoarseEffect::NORMAL, is_conditional}};
+  } else {
+    return {{CoarseEffect::ABNORMAL, is_conditional}};
+  }
```

Comment:
I would extract that check into some helper (`flow_normality( x, y ) -> CoarseEffect`?